docs/src/appendix/cheatcodes/l1_handler_call.md

```diff
@@ -0,0 +1,73 @@
+# `l1_handler_call`
+
+> `fn l1_handler_call(prepared_l1_handler: PreparedL1Handler)`
+
+Calls a `#[l1_handler]` function to mock a
+[message](https://docs.starknet.io/documentation/architecture_and_concepts/L1-L2_Communication/messaging-mechanism/)
+arriving from Ethereum.
+
+```rust
+struct PreparedL1Handler {
+    contract_address: ContractAddress,
+    selector: felt252,
+    from_address: felt252,
+    payload: Span::<felt252>,
+}
+```
+
+where:
+
+- `contract_address` - target contract address
+- `selector` - selector of a `#[l1_handler]` function
+- `from_address` - ethereum address which sends the message
+- `payload` - message payload
+
+It is important to note that the `from_address` must always be
+checked as any L1 contract can call any L2 contract.
+
+The payload contains any Cairo data structure that can be serialized with
+[Serde](https://book.cairo-lang.org/appendix-03-derivable-traits.html?highlight=serde#serializing-with-serde).
+
+For contract implementation:
+
+```rust
+// ...
+#[storage]
+struct Storage {
+    l1_caller: felt252,
+    //...
+}
+
+//...
+
+#[l1_handler]
+fn process_l1_message(ref self: ContractState, from_address: felt252, data: Span<felt252>) {
+    assert(from_address == self.l1_caller.read(), 'Unauthorized l1 caller');
+}
+// ...
+```
+
+We can use `l1_handler_call` to call a `l1_handler`, which is not accessible using a dispatcher:
+
+```rust
+use snforge_std::l1_handler_call;
+
+#[test]
+fn test_l1_handler_call() {
+    // ...
```

Comment:
I do like this idea of having something we can build, and then execute, instead of the prepare fashion. :+1: I'll go that way thanks a lot @MaksymilianDemitraszek.

About the name, as commented to @Radinyn I am not sure `mock` is that accurate IMO as the call to the function is done. But it effectively can be more consistent in the naming convention you already have.

Reply:
I agree Mock is not accurate what to you think about `trigger`?

Reply:
`trigger` sounds nice. I was also thinking about `l1_handler_invoke` because it is what we do and Starkware documents it like this.

`call` sounds too much like a regular function so yes it should be discarded.

Reply:
> Maybe we could create a more extensible interface here 🤔 What do you thing about something like this
> 
> ```
> MockL1Call::build(contract_address, function_name, payload).execute()
> ```
> 
> and if somebody wants to include additional configuration
> 
> ```
> let l1mock = MockL1Call::build(contract_address, function_name, payload).execute()
> l1mock.from_address = 0x123123
> l1mock.paid_fee_on_l1 = 0x123123
> ```
> 
> What do you think?

After doing some tests, I am not sure how to use this approach.

We almost always want to set `from_address` for instance. Which mean I'll always have 2 lines of code to init the struct.

```rust
let l1_handler = L1HandlerInvoker::build(contract_address, selector_name, payload);
l1_handler.from_address = 0x1234;
l1_handler.invoke();
```

However, what's is very interesting in your approach is that when we test a contract, we don't want to always repeat the `contract_address` when changing the arguments to see how the contract behaves. Same for the selector name. We want to play with `payload` and `from_address` mostly, to ensure our `l1_handler` implementation is correctly setup.

What about this:

```rust
// Contract 1

#[l1_handler]
fn accept_l1_data(from_address: felt252, value1: felt252, value2: felt252) {
    assert(from_address == 0x123, 'Invalid L1 address');
    assert(value1 != value2, 'Invalid values');
}
```

Then we can:

```rust
// Test function

let contract_address = deploy(...);

// Build the reusable struct.
let l1_handler = L1Handler::build(contract_address, 'accept_l1_data');

// Invoke with given parameters (contract_address, payload, fees).
l1_handler.invoke(0x123, @array![0x11, 0x22], 0); // << will pass
l1_handler.invoke(0x999, @array![0x1, 0x1], 0); // << will revert
```
What's your take on this one?

Also @Radinyn and @MaksymilianDemitraszek for the naming, what do you think about the convention given above:
`L1Handler` struct to build the targetted handler.
`.invoke` to actually fire the TX and the underlying execution flow.